compiler/rustc_metadata/src/locator.rs

```diff
@@ -394,46 +398,97 @@ impl<'a> CrateLocator<'a> {
+
+            // njn: factor out more similarities?
+
+            /*
+            let mut s = String::new();
+            s.push_str(&format!("find {} in {}\n", rlib_prefix, search_path.dir.display()));
+            for rlib in rlibs {
+                s.push_str(&format!("- {}\n", rlib.file_name_str));
+            }
+            s.push_str(&format!("linear {:?}\n", linear_matches));
+            s.push_str(&format!("binary {:?}\n", binary_matches));
+            eprint!("***\n{}***\n", s);
+            */
+
+            // njn: rename
+            // njn: kind isn't needed as an arg
+            let mut add_candidate = |table: &mut FxHashMap<_, _>, path: &PathBuf| {
+                let path = fs::canonicalize(&path).unwrap_or_else(|_| path.clone());
+                if !seen_paths.contains(&path) {
+                    seen_paths.insert(path.clone());
+                    table.insert(path, search_path.kind);
+                };
             };
-            let (hash, found_kind) = if file.starts_with(&rlib_prefix) && file.ends_with(".rlib") {
-                (&file[(rlib_prefix.len())..(file.len() - ".rlib".len())], CrateFlavor::Rlib)
-            } else if file.starts_with(&rlib_prefix) && file.ends_with(".rmeta") {
-                (&file[(rlib_prefix.len())..(file.len() - ".rmeta".len())], CrateFlavor::Rmeta)
-            } else if file.starts_with(&dylib_prefix) && file.ends_with(&self.target.dll_suffix) {
-                (
-                    &file[(dylib_prefix.len())..(file.len() - self.target.dll_suffix.len())],
-                    CrateFlavor::Dylib,
-                )
-            } else {
-                if file.starts_with(&staticlib_prefix)
-                    && file.ends_with(&self.target.staticlib_suffix)
+
+            // .rlib files
+            let rlibs = &search_path.rlibs;
+            // njn: explain this
+            let i = rlibs.partition_point(|spf| spf.file_name_str < rlib_prefix);
+            let mut j = i;
+            while let Some(true) =
+                rlibs.get(j).map(|spf| spf.file_name_str.starts_with(&rlib_prefix))
+            {
+                j += 1;
+            }
+
+            // Process the matches.
+            for spf in &rlibs[i..j] {
+                let name = &spf.file_name_str;
+                assert!(name.ends_with(".rlib"));
+                let hash = &name[(rlib_prefix.len())..(name.len() - ".rlib".len())];
+                info!("lib candidate: {}", spf.path.display());
+
+                let (rlibs, _rmetas, _dylibs) = candidates.entry(hash.to_string()).or_default();
+                add_candidate(rlibs, &spf.path);
+            }
+
+            // .rmeta files
+            let rmetas = &search_path.rmetas;
+            // njn: explain this
+            let i = rmetas.partition_point(|spf| spf.file_name_str < rmeta_prefix);
+            let mut j = i;
+            while let Some(true) =
+                rmetas.get(j).map(|spf| spf.file_name_str.starts_with(&rmeta_prefix))
```

Comment:
I think this is done more understandably via `let j = rmetas.iter().filter(|spf| ...).count();`.

Reply:
That would go back to linear search, no? The whole point of the last commit is to change linear search to binary search.

Reply:
That code was doing a linear search. It was incrementing `j` in steps of 1. My suggested replacement would not have exited early though while your code did, so your code was a bit faster, even though it wasn't a binary search.

Reply:
`partition_point` does binary search, then there was a short linear search (usually length 0 or 1) involving `j` after that. But it's moot anyway because it ended up not helping performance.